--- a/version.py
+++ b/version.py
@@ -95,7 +36,7 @@ def llvm_cmake_version(llvm_dir):

 def main(action, llvm_dir):
     if action == 'wasi-sdk':
-        print(git_version())
+        print("nix")
     elif action == 'llvm':
         major, minor, path = llvm_cmake_version(llvm_dir)
         print(f'{major}.{minor}.{path}')
@@ -103,12 +44,12 @@ def main(action, llvm_dir):
         major, _, _ = llvm_cmake_version(llvm_dir)
         print(major)
     elif action == 'dump':
-        print(git_version())
-        print(f'wasi-libc: {git_commit("src/wasi-libc")}')
-        print(f'llvm: {git_commit(llvm_dir)}')
+        print("nix")
+        print(f'wasi-libc: nix')
+        print(f'llvm: nix')
         major, minor, path = llvm_cmake_version(llvm_dir)
         print(f'llvm-version: {major}.{minor}.{path}')
-        print(f'config: {git_commit("src/config")}')
+        print(f'config: nix')


 if __name__ == '__main__':
